src/Toml/BiMap.hs

```diff
@@ -51,245 +55,270 @@ module Toml.BiMap
-prism review preview = BiMap preview (Just . review)
+prism :: (field -> object) -> (object -> Either error field) -> BiMap error object field
+prism review preview = BiMap preview (Right . review)
+
+----------------------------------------------------------------------------
+-- BiMap error types
+----------------------------------------------------------------------------
+
+-- | Abstraction on BiMap with error.
+type TomlBiMap = BiMap TomlBiMapError
+
+prettyBiMapError :: TomlBiMapError -> Text
+prettyBiMapError = \case
+    WrongConstructor expect actual ->
+        "Invalid type constrictor. Expected " <> expect <> " Actual constructor: " <> actual
+    WrongValue (MatchError expect actual) ->
+        "Invalid value constructor. Expected " <> expect <> " Actual value: " <> toTxt actual
+    ArbitraryError text  -> "Invalid value: " <> text
 
 ----------------------------------------------------------------------------
 -- General purpose bimaps
 ----------------------------------------------------------------------------
 
 -- | Bimap for 'Either' and its left type
-_Left :: BiMap (Either l r) l
-_Left = prism Left (either Just (const Nothing))
+_Left :: (Show l, Show r) => TomlBiMap (Either l r) l
+_Left = prism Left $ \case
+    Left l -> Right l
+    x -> Left $ WrongConstructor "Left" (toTxt x)
 
 -- | Bimap for 'Either' and its right type
-_Right :: BiMap (Either l r) r
-_Right = prism Right (either (const Nothing) Just)
+_Right :: (Show l, Show r) => TomlBiMap (Either l r) r
+_Right = prism Right $ \case
+    Right r -> Right r
+    x -> Left $ WrongConstructor "Right" (toTxt x)
 
 -- | Bimap for 'Maybe'
-_Just :: BiMap (Maybe a) a
-_Just = prism Just id
+_Just :: Show r => TomlBiMap (Maybe r) r
+_Just = prism Just $ \case
+    Just r -> Right r
+    x -> Left $ WrongConstructor "Just" (toTxt x)
 
 ----------------------------------------------------------------------------
 --  BiMaps for value
 ----------------------------------------------------------------------------
 
 -- | Creates prism for 'AnyValue'.
-mkAnyValueBiMap :: (forall t . Value t -> Maybe a)
+mkAnyValueBiMap :: (forall t . Value t -> Either TomlBiMapError a)
                 -> (a -> Value tag)
-                -> BiMap a AnyValue
-mkAnyValueBiMap matchValue toValue =
-    BiMap (Just . AnyValue . toValue) (\(AnyValue value) -> matchValue value)
+                -> TomlBiMap a AnyValue
+mkAnyValueBiMap matchValue toValue = BiMap (Right . AnyValue . toValue)
+    (\(AnyValue value) -> either (\_ -> leftWrongValue value) Right $ matchValue value)
 
 -- | Creates bimap for 'Text' to 'AnyValue' with custom functions
-_TextBy :: (a -> Text) -> (Text -> Maybe a) -> BiMap a AnyValue
-_TextBy toText parseText =
-  mkAnyValueBiMap (matchText >=> parseText) (Text . toText)
+_TextBy :: (a -> Text) -> (Text -> Either TomlBiMapError a) -> TomlBiMap a AnyValue
+_TextBy toText parseText = mkAnyValueBiMap (matchText >=> parseText) (Text . toText)
```

Comment:
I don't expect users to pass parsing function from `Text` to `Either TomlBiMapError` here 🙂 

It's more convenient if this function has type: `Text -> Either Text a)` and you can wrap this text into `TomlBiMapError`

Reply:
I think even expecting a `Text -> Either Text a` from users is a lot. Can we keep it `Text -> Maybe a`?

Reply:
Well, the idea here is that in most cases users should pass here some function that outputs proper error in case of failed parsing. At least it should be `readEither`. If you have function of type `Text -> Maybe a` then you can easily convert it to `Text -> Either Text a` by providing some error message for `Nothing`. It should be users responsibility to tell proper error message.